source/application/DbgHelp/DbgHelpStackWalkCallbackPrivate.cpp

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2016- Hourglass Resurrection Team
+ * Hourglass Resurrection is licensed under GPL v2.
+ * Refer to the file COPYING.txt in the project root.
+ */
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+#include <sstream>
+#include <string>
+
+#include <atlbase.h>
+#include <atlcom.h>
+
+#include "application/logging.h"
+#include "DbgHelpPrivate.h"
+#include "DbgHelpStackWalkCallbackPrivate.h"
+
+DbgHelpStackWalkCallbackPrivate::DbgHelpStackWalkCallbackPrivate(HANDLE process, IDiaStackFrame* frame, const DbgHelpPrivate::ModuleData* mod_info) :
+    m_process(process),
+    m_frame(frame),
+    m_mod_info(mod_info),
+    m_unsure(0),
+    m_params_enumerated(false)
+{
+}
+
+// TODO: Split module name and PC
+std::wstring DbgHelpStackWalkCallbackPrivate::GetModuleName()
+{
+    std::wstringstream name;
+    if (m_mod_info == nullptr)
+    {
+        name << L"?";
+    }
+    else
+    {
+        name << m_mod_info->m_module_name;
+    }
+    name << L"!0x" << std::hex << GetProgramCounter();
+    return name.str();
+}
+
+std::wstring DbgHelpStackWalkCallbackPrivate::GetFunctionName()
+{
+    if (m_mod_info == nullptr)
+    {
+        return L"?";
+    }
+    std::wstring function_name = L"?";
+    if (m_mod_info->m_module_symbol_session == nullptr)
+    {
+        auto it = m_mod_info->m_module_exports_table.lower_bound(GetProgramCounter());
+        if (it != m_mod_info->m_module_exports_table.begin() && (--it) != m_mod_info->m_module_exports_table.begin())
+        {
+            function_name = it->second;
+        }
+        return function_name;
+    }
+    auto symbol = GetFunctionSymbol();
+    
+    BSTR name = nullptr;
+    if (symbol != nullptr && symbol->get_name(&name) == S_OK)
+    {
+        function_name = name;
+        SysFreeString(name);
+    }
+    return function_name;
+}
+
+// TODO: RFC
+std::wstring DbgHelpStackWalkCallbackPrivate::GetFunctionParameters()
+{
+    if (m_mod_info == nullptr)
+    {
+        return L"?";
+    }
+
+    ULONGLONG stackframe_base = 0;
+    bool can_get_values = (m_frame->get_base(&stackframe_base) == S_OK);
+
+    DWORD param_count = GetParameterCount();
+    std::wstring params = L"(";
+    //if (m_mod_info->m_module_symbol_session == nullptr)
+    {
+        /*
+         * We don't know the parameter names, nor types, nor the actual count, but we have a guess.
+         * We'll also assume that there's no more than 26 parameters to a function.

```

Comment:
Another question is how useful this can be at all? Perhaps not returning anything in case of unknown parameters is the best option? Maybe some way to manually set the parameter count and types through the UI, so through arguments to this (since if this is ever useful it will imply the person will go look up the actual parameter count and types for the function)?

Otherwise, distance to the next frame could provide a better approximation.